### PR TITLE
Add SourceLink support to package

### PR DESCRIPTION
**Pull request by matracey**
## New features and updates

[`522be95`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/522be95) add debug symbols and explicit packable configuration to `src/Directory.build.props`
[`f1fd5d7`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/f1fd5d7) add explicit `SourceLink.GitHub` package reference
[`31bd648`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/31bd648) add msbuild properties for `SourceLink`
[`e72126c`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/e72126c) upgrade packages to latest versions

## Project structure enhancements

[`558b8cb`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/558b8cb) move `sln` file up to project root
[`f616e41`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/f616e41) add missing loose items to `sln` file
[`f4ef2d5`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/f4ef2d5) add recommended VS Code extensions to `extensions.json`
[`8cecd49`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/8cecd49) update `gitignore` using `dotnet new gitignore`
[`7b1e292`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/7b1e292) add `global.json` with SDK version configuration
[`d554044`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/d554044) add `IsTestProject` property to `Directory.Build.props` for test folder
[`5ea3667`](https://github.com/weichch/system-text-json-jsondiffpatch/pull/42/commits/5ea3667) update `AssemblyVersion` and `FileVersion` to use `JsonDiffPatchPackageVersion`

### CODE REVIEW

**Comment by matracey** [comment]
@weichch Are you still maintaining this package? Happy to take it over if not.

File: SystemTextJson.JsonDiffPatch.sln [hunk 1]
```diff
@@ -0,0 +1,138 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (weichch) [inline SystemTextJson.JsonDiffPatch.sln:2]: Why moved solution file here?

File: src/SystemTextJson.JsonDiffPatch.MSTest/SystemTextJson.JsonDiffPatch.MSTest.csproj [hunk 2]
```diff
@@ -1,21 +1,25 @@
...
-  </ItemGroup>
+    <ItemGroup>
+        <ProjectReference Include="..\SystemTextJson.JsonDiffPatch\SystemTextJson.JsonDiffPatch.csproj" />
+    </ItemGroup>
 
-  <ItemGroup>
-    <PackageReference Include="MSTest.TestFramework" Version="2.1.2" />
-  </ItemGroup>
+    <ItemGroup>
+        <PackageReference Include="Microsoft.SourceLink.GitHub" Version="8.0.0">
```
Comment (weichch) [inline src/SystemTextJson.JsonDiffPatch.MSTest/SystemTextJson.JsonDiffPatch.MSTest.csproj:18]: What's the purpose of adding this?

File: ReleaseNotes.md [hunk 3]
```diff
@@ -1,5 +1,24 @@
...
+- Added `global.json` with SDK version configuration
+- Moved solution file to project root
+- Added recommended VS Code extensions to `extensions.json`
+- Added VS Code settings for formatting and file associations
+- Added tool manifest using `dotnet new tool-manifest`
+- Added `dotnet-format` to tool manifest
+- Added ReSharper/Rider config
+- Formatted `README.md`, `Benchmark.md`, `ReleaseNotes.md`, `Directory.Build.props` files, `csproj` files, and `test/Examples` json files
+- Ran `dotnet format` for error, warn, and info severity
+- Moved test `TargetFrameworks` to `Directory.Build.props` for test folder
```
Comment (weichch) [inline ReleaseNotes.md:20]: I appreciate all the valuable changes you're proposing, but can we have separate contributions (backed by issues created first please) to address the following:
- adding dotnet-format
- adding support for VS Code contributors

Those non-functional changes wouldn't be necessary listed in the release notes as the release notes are publicly visible to the package users. Please only keep them in the description of the PR.

File: ReleaseNotes.md [hunk 4]
```diff
@@ -1,5 +1,24 @@
 # Release Notes
 
+## 2.0.0
+
+- Add support for `net462`, `net7.0` and `net8.0`
+- **[BREAKING CHANGE]** Drop support for `net461`, `netcoreapp3.1` and `net5.0`, all of which are EOL
```
Comment (weichch) [inline ReleaseNotes.md:6]: I appreciate we're making a breaking change to remove some of the EOL packages, but I'd like to see the targeting frameworks align with Microsoft's `System.Text.Json`, which currently supports `netstandard2.0`, `net6.0`, `net7.0`, `net8.0` and `net462`. 

Because I accidentally added `netstandard2.1` from the very beginning (for the purpose of using nullable reference etc), we also need to keep that and if it really bothers, we can consider it in another PR to consolidate it.

We won't be able to remove `net462` from the targeting framework list even though it has been EOL because that will force our .NET Framework users to recompile against `netstandard2.0` which is not fully supported until .NET Framework 4.7.2.

**Review by weichch — COMMENTED** [review]
Hi @matracey Thanks again for taking effort to contribute. Like I mentioned in the comments, can we please track the repository house-keeping changes in a separate issue? 

It's a little hard to follow the actual changes in this PR with more than 100 files changed due to reformatting.

**Comment by weichch** [comment]
I cherry picked some of the changes in this PR into #43 as co-authored commit. Let's keep this PR open for addressing VS Code support if you wish.

I also found Microsoft has added two new methods in .NET 8 `JsonNode.DeepEquals` and `JsonNode.DeepClone`. Those methods are conflicting with the extension methods in this package as per mentioned in #4, and I will take a look at how to address those. Meanwhile, those methods don't seem to exist in .NET 6 and .NET 7.

File: src/SystemTextJson.JsonDiffPatch.MSTest/SystemTextJson.JsonDiffPatch.MSTest.csproj (hunk 2 above)
Comment (matracey) [reply src/SystemTextJson.JsonDiffPatch.MSTest/SystemTextJson.JsonDiffPatch.MSTest.csproj]: I added SourceLink because it provides a great source debugging experience for end users. It provides a lot of debug-time assistance, allowing to step into the source code of the compiled NuGet package by pulling the source from GitHub, rather than trying to use decompilation. [SourceLink is supported by Microsoft and included in the .NET SDK](https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink) for a variety of common source control providers, including GitHub, and adding it provides a quick win for a much improved package consumer experience.

File: SystemTextJson.JsonDiffPatch.sln (hunk 1 above)
Comment (matracey) [reply SystemTextJson.JsonDiffPatch.sln:2]: I thought it makes more sense to have the sln file at the root, since the src folder will contain the source for the library itself, and the test folder will contain the source code for the tests.

IMO, it's easier to find the sln file in the root, since many open source .NET repos follow this paradigm, and it makes more sense to reference like `src\SystemTextJson.JsonDiffPatch\SystemTextJson.JsonDiffPatch.csproj` & `test\SystemTextJson.JsonDiffPatch.UnitTests\SystemTextJson.JsonDiffPatch.UnitTests.csproj` rather than `SystemTextJson.JsonDiffPatch\SystemTextJson.JsonDiffPatch.csproj` & `..\test\SystemTextJson.JsonDiffPatch.UnitTests\SystemTextJson.JsonDiffPatch.UnitTests.csproj`.

File: ReleaseNotes.md (hunk 4 above)
Comment (matracey) [reply ReleaseNotes.md]: This is a fair point. I thought it might be a (mostly) transparent change, but you're right that there could be some edge cases where the change in support could break things for some users.

File: ReleaseNotes.md (hunk 3 above)
Comment (matracey) [reply ReleaseNotes.md]: This is completely fair. I've stripped back the changes in this PR to be much more focused on specifically adding [SourceLink](https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink) and I will raise a separate PR to add `dotnet-format` along with the default .NET editorconfig.
